File: golang/graphs/graphs.go
package graphs

import "fmt"

type Node struct {
	data       interface{}
	neighbours []*Node
	visited    bool
}

func NewNode(data interface{}) *Node {
	return &Node{
		data:       data,
		neighbours: []*Node{},
	}
}

func (node *Node) Connect(neighbour *Node) {
	node.neighbours = append(node.neighbours, neighbour)
}

func (node *Node) String() string {
	if node.visited {
		return ""
	}
	var s string
	node.visited = true
	s += fmt.Sprintf("%v\n", node.data)
	for _, n := range node.neighbours {
		s += fmt.Sprintf("%v -- %v\n", node.data, n.data)
		s += fmt.Sprintf("%v", n)
	}
	return s
}

func (node *Node) DotDump() string {
	var s string
	s += fmt.Sprintln("graph {")
	s += fmt.Sprintf("%v", node)
	s += fmt.Sprintln("}")
	return s
}

File: Cpp/graphs.cpp
#include "graphs.h"

Node::Node(const void *data) : data{data}, visited{false} {}

Node::~Node() = default;

void Node::Connect(std::shared_ptr<Node> node) { neighbours.push_back(node); }

std::shared_ptr<Node> newNode(const void *data) {
  return std::shared_ptr<Node>(new Node{data});
}

void dot_dump(std::ostream &os, const std::shared_ptr<Node> &node) {
  os << "digraph {\n" << node << "}\n";
}

std::ostream &
operator<<(std::ostream &os, std::shared_ptr<Node> node) {
  if (node->visited) {
    return os;
  }

  node->visited = true;
  os << *(const char *)node->data << "\n";
  for (auto &n : node->neighbours) {
    os << *(const char *)node->data << " -- " << *(const char *)n->data << "\n";
    os << n;
  }
  return os;
}
